alf/config_util.py

```diff
@@ -0,0 +1,480 @@
+
+    def set_used(self):
+        self._used = True
+
+    def is_used(self):
+        return self._used
+
+
+# _CONF_TREE is a suffix tree. For a name such as "abc.def.ghi", the corresponding
+# node can be found using _CONF_TREE['ghi']['def']['abc']
+_CONF_TREE = {}
+
+
+def config1(config_name, value, replacing_existing_config=False):
+    """Set one configurable value.
+
+    Args:
+        config_name (str): name of the config
+        value (any): value of the config
+        replacing_existing_config (bool): whether to replace with the provided
+            new value if the value has be configured before. A warning will be
+            generated if this is False and the value has been set previously.
+
+    """
+    tree = _CONF_TREE
+    path = config_name.split('.')
+    for name in reversed(path):
+        if not isinstance(tree, dict) or name not in tree:
+            raise ValueError("Cannot find config name %s" % config_name)
+        tree = tree[name]
+
+    leaves = _get_all_leaves(tree)
+    if len(leaves) > 1:
+        # only show atmost 3 ambiguous choices
+        leaves = leaves[:3]
+        names = [name + '.' + config_name for name, node in leaves]
+        raise ValueError("config name '%s' is ambiguous. There are %s" %
+                         (config_name, names))
+
+    assert len(leaves) == 1
+    config_node = leaves[0][1]
+    if config_node.is_used():
+        raise ValueError(
+            "Config '%s' has already been used. You should config "
+            "its value before using it." % config_name)
+    if not replacing_existing_config and config_node.is_configured():
+        logging.warning(
+            "The value of config '%s' has been configured. The new "
+            "config will be ignored" % config_name)
+    else:
+        config_node.set_value(value)
+
+
+def _make_config(signature, whitelist, blacklist):
+    """
+    Args:
+        signature (inspec.Signature):
+    Returns:
+        dict: name => _Config
+    """
+    configs = {}
+    for name, param in signature.parameters.items():
+        if param.kind in (inspect.Parameter.VAR_POSITIONAL,
+                          inspect.Parameter.VAR_KEYWORD):
+            continue
+        if ((not blacklist and not whitelist)
+                or (whitelist and name in whitelist)
+                or (blacklist and name not in blacklist)):
+            config = _Config()
+            configs[name] = config
+            if param.default is not inspect.Parameter.empty:
+                config.set_default_value(param.default)
+
+    return configs
+
+
+def _add_to_conf_tree(module_path, func_name, para_name, node):
+    """
+    Args:
+        module_path (list[str]):
+        node (_Config):
+    """
+
+    tree = _CONF_TREE
+    path = module_path + func_name.split('.') + [para_name]
+    names = []
+    for name in reversed(path[1:]):
+        if not isinstance(tree, dict):
+            raise ValueError("'%s' conflicts with existing config name '%s' "
```

Comment:
This error message is confusing. Can be `"%s path is not valid; only found a partial path up to %s" % ('.'join(path), '.'.join(reversed(names)))`

Reply:
I'm actually not sure about the intended usage here. Suppose we have a function called "B.a", and another function "A.B.a". So basically if I add "A.B.a" first and then "B.a", there will be no error. But I can't do in the reversed order?